HttT Fix errors and warnings, and misc. improvements

GregoryLundberg opened the pull request:
[REWRITE]

This is a result of a careful code reading and walk-through in an attempt to find and fix as many issues as possible.

Items marked 'Fix bug' are the most important as they correct obvious errors.

Items marked 'Workaround' avoid an error in the game engine and should be reversed when corrected.

All other items are discretionary. They represent only my opinion of an improvement. Most notable of these is the addition of a sub-quest for to assist the wose in scenario 20b ("Underground Channels").

--

To shorten this stream, I have gone through and deleted a bunch of my comments which have been addressed elsewhere or are no longer germane.

GregoryLundberg commented:
After play-testing, these are my notes on messages seen on stderr:

Console messages while play-testing 'Heir to the Throne':

    libpng warning: iCCP: known incorrect sRGB profile

Appears consistently upon victory when Seimus appears. Seems more random elsewhere.



Heir to the Throne, scenario 7, "Crossroads"

    error engine: Trying to add Kalenz - Kalenz - 2 (2,13) over Delfador - Delfador - 2 (2,10).
    error engine: The new unit was assigned underlying_id=100 to prevent duplicate id conflicts.

Delfador was unstored, and Kalenz created, during the victory processing in scenario 6, "The Siege of Elensefar". During scenario start-up they are both recalled. Inspecting Kalenz in S06 shows a large underlying_id, while Delfador is 2. So why does this error appear?



Heir to the Throne, scenario 17, "Scepter of Fire"

    error scripting/lua: stoi
    stack traceback:
	    [C]: in function 'get_locations'
	    lua/wml-tags.lua:584: in function 'cmd'
	    lua/wml-utils.lua:137: in function 'handle_event_commands'
	    lua/wml-flow.lua:6: in function <lua/wml-flow.lua:5>

Consistently appears both on the console log (stderr) and on screen.

    error ai/actions: Error #2007 (E_NOT_REACHED_DESTINATION) in full move by side 4 from location 38,50 to location 37,50

This is a leader trying to move to a hex where it just recruited a unit into. Just next door, in castle.

GregoryLundberg commented:
About the re-assignment of the underlying_id message: I don't know what caused it, here, but, while working out how to store and unstore the Merman for S18/S19c I was able to cause it by accidentially storing Konrad and his entire recall list instead of storing just the Merman I wanted. Not only did that cause a similar message, it actually created TWO copies of Konrad, and no other units; and the death of either was sufficient for the failure objective. Not a bug, per se, but it was sort of fun. I changed one Konrad's side and had Konrad kill Konrad, for grinz, before fixing my mistake.

gfgtdf commented:
> Delfador was unstored, and Kalenz created, during the victory processing in scenario 6, "The Siege of Elensefar". During scenario start-up they are both recalled. Inspecting Kalenz in S06 shows a large underlying_id, while Delfador is 2. So why does this error appear?

The mostlikeley happens because the game considers "victory" to be a unsave event (In mp it's possible that one client wins while the other looses, and in that case the victory evnet is only fired on the winning side) so its doesnt like creating units during a victory event. There is the 'scenario end' even which is synced but this gets also fired in case of a defeat, not it might be unsuited for this case, will think about this.

CelticMinstrel commented:
What's wrong with simply syncing the victory event? Yes, it'll only be fired on victorious clients, but there should be a way to transmit any results of it to the other clients.

CelticMinstrel commented on data/campaigns/Heir_To_The_Throne/scenarios/04_The_Bay_of_Pearls.cfg:
```diff
@@ -1,4 +1,77 @@
+    "Mermaid Initiate"
+#enddef
+
+#define UnitSpeaks RoleName UnitList SpokenLine
```
This macro seems to utterly defeat the point of using `[role]` in the first place. If I'm not mistaken, you could simply delete lines 53-56 and it would still work. Furthermore, in that case the `[if]` conditional is meaningless since a `[message]` is automatically and silently skipped if no speaker is found.

More generally, I'm not convinced this is the best solution. Did you consider adding a death event to reassign essential roles? Something like the following:

```wml
[event]
  name=die
  [filter]
    role=some_role_name
  [/filter]
  [role]
    role=some_role_name
    side=1
    type=whatever
  [/role]
[/event]
```

I guess you'd need two of them, one for the thankful merfolk and one for the advisor.

CelticMinstrel commented on data/campaigns/Heir_To_The_Throne/scenarios/05a_Muff_Malal_Peninsula.cfg:
```diff
@@ -1,4 +1,40 @@
 #textdomain wesnoth-httt
+
+#define ADVISORS
```
When you have something duplicated in two or more scenarios, it's probably better to move it into one of the general macro files under _utils/_ or _macros/_ (whichever HttT calls it).

gfgtdf commented:
Not really sure what you mean, beeing executed on all clients (on the same way) is exactly what synced means.

GregoryLundberg commented:
The macro names are the same, but the units listed are differ.

I suppose I should have used different macro names to clue you in to the subtle differences.

The [role] tag searches in the order of the type= attribute. So I am using that to select the higher (or lower) levels. I know it's verbose and sorta disgusting, but selecting on race= would just pick one (yes, that would be deterministic, but not meaningful).

As to using a shared marco .. I am still working out the use cases. I hope, at some point, to be able to write a small set of general-purpose macros and then use them throughout the entire set of mainline campaigns. It seems, so far, that every day I am finding another use case. sigh.

GregoryLundberg commented:
I read the comments on the Wiki about syncing events and saw the rationale about the victory and defeat conditions.

As I read it, the event in question is in S06, "enemies defeated". During that event, Delfador is unstored and Kalenz is created using [unit]. The scenario then ends and both Delfador and Kalenz are recalled at the start of S07.

I was working on something similar on AToTB but the leader there was never unstored. A new unit was, however, created with [unit]. In the case I was hitting there, no errors appeared until in the next scenario, I recalled the new unit. And **THEN** the **ONLY** time it generated the warning was when I was 'playing through' without stopping or loading. If I loaded the second scenario (Back to Start, etc) the error does **not** appear.

CelticMinstrel commented:
gfgtdf, I just mean that the victory/defeat events should probably be synched.

ln-zookeeper commented:
How big part of the motivation for all these role-related things is to avoid the "no matching unit found" errors? Because it turns out that those errors shouldn't appear in the first place, someone just added them for unknown reasons.

Of course there's nothing wrong in adding failsafes to avoid dialogue getting skipped, but suppressing the actual error message shouldn't factor into it.

gfgtdf commented:
If not finding a unit in [role] usualy imples an error later then it makes sense to give such a warning, whether this is the case should be determined by looking at example cases, for example here in MermanOffersHelp and AdvisorSpeaks this is not the case but if the advisor unit is lacking in 05a, the next dialog doesn't really make sense,

I personnaly suport vultraz idea of addin [then] [else] to [role] so that we'D onyl give the warning oin case no [else] is given

GregoryLundberg commented:
As I see it, the current implementation of [role] leads to down-stream edge-case problems.

For instance, unlike [have_unit], role searches first on the board, then in the recall list.

Consider looking for a Horseman for a speaking part. If you have Horsemen on the board, [role] finds them, assigns the role to one, and the lines are spoken. If you have no Horsemen at all, the [role] displays an on-screen error. But, if you have a Horseman in the recall list, no error appears, the role is assigned to the not-recalled Horseman, no errors occur, and the spoken lines are silently skipped.

If that skipped line is a color comment, you probably don't miss it.

But consider the poor player trying to make sense of:

    Leader: Yes, that's right
    Leader: OK, let's go!

Instead of:

    Adviser: To win the day we need to kill the Lich to the north, first.
    Leader: Yes, that's right.
    Adviser: But, we cannot head straight north. We need to approach through the woods on the west.
    Leader: OK, let's go!

And all because the Adviser is a Horseman but all our Horsemen are in the recall list, none on the board.

So, it's not simply adding [else] or some other on-not-found clause. We also need 'search_recall_list=no' instead of the default 'yes' and for that we need search_recall_list added as an allowed attribute of [role], with the default switched to 'yes'

AI0867 commented:
I suspect the best place to error out would be in the actual [message] tag. An ignore_missing key would need to be added for the cases where units can legitemately be missing, which does occur.

GregoryLundberg commented:
Well, that's assuming the only case we want to fix is the [message] case.

I started doing this because of stilted conversations .. it's pretty obvious.  Then I realized that it was also the WML role not found errors.

Now, I simply grep for 'role=' because there are other use cases, but they're all related to [role] either failing or succeeding incorrectly.

CelticMinstrel commented:
I see, I wasn't aware of the problem with `[role]` assigning a recall list unit. A `search_recall_list=no` option is one idea, and `recall=yes` could be another idea (to automatically recall the unit if it's not on the map).

I'm divided about the "no unit found for role" warning. On the one hand, it seems like it aided you in finding many of these problems in conversation (unless I'm mistaken?), but on the other hand it's a warning that in some (many?) cases is something not to be worried about (and thus could maybe be removed). I considered adding `optional=yes` to `[role]` simply to suppress the warning. I also considered lowering the priority so that it doesn't appear at the default logging level. Vultraz came up with the idea of adding `[else]` to `[role]`, so obviously the presence of that would suppress the warning.

GregoryLundberg commented:
The problem is that the message appears on-screen. So it's not silently dropping messages for non-existent units, it's informing the player the unit doesn't exist.

That's what started me down this road .. everything seemed normal, yet there was a message on the top-right corner of my screen while playing.

Remember, just because the typical use is for messages, there are other uses. Anywhere you might use a filter with a role= is a potential source of problems if the presumption is the player will always have, will never have, will have in the recall list, etc.

CelticMinstrel commented:
It's the `[role]` tag that prints the message, as I understand. If the role gets assigned but then dies before it's time for it to speak, you won't get anything displayed.

While your work here is (at least in part) quite useful, I think in some cases (when the dialog flows fine without that speaker) it may be unnecessary, and for those cases changing how `[role]` works is actually better.

GregoryLundberg commented:
I agree, changing how [role] works would be a better solution.

The question is what it should change to, and when it would change. I, certainly, am not up to the task in any reasonable time-frame, or I'd volunteer.

I am also not sure we truly understand all the ways [role] is currently used which cause failures.

At present, I think the following _might_ get us where we'd need to be:

1) Add a new attribute, search_recall_list. This allows the WML to control whether [role] searches the recall list. Default: yes.

2) Add a new allowed value for search_recall_list (also for have_unit): only. This tells [role] to skip consideration of units on the board and only consider units in the recall list.

3) Modify [role] to behave as both an imperative and a conditional tag. This would allow existing usage and add the ability to use it in both [if] and [while] and combined (possibly with other [role] tags) using [or] and [and].

CelticMinstrel commented:
Isn't your third point the same as `[have_unit]role=blah[/have_unit]`?

GregoryLundberg commented:
But [role] has a side-effect of setting the role if none exists. Which means combining two [role] in an implicit [and] would have different effect if we short-curcuit .. something to consider.

(which is why some languages short-circuit AND and OR and offer ANDTHEN  and ORELSE to not short circuit ...)

CelticMinstrel commented:
So wait, you're suggesting a conditional tag with side-effects? I'm not sure I like this idea, honestly.

GregoryLundberg commented:
Well it's conditionally failing.

And conditionally incorrectly succeeding.

I saw your commit to remove the message. All that does is hide that it's not working as expected when it unconditionally fails. Someone (me?) still needs to run through all the scenarios and check that every use of [role] and role= operate as expected.

The lack of a failure message is not indicative of success.

GregoryLundberg commented:
Let's take the simplest case.

Which is more correct / preferred:

    [role]
        role=rider
        type=Horseman
    [/role]
    [if]
        [have_unit]
            role=rider
        [/have_unit]
        [then]
        .
        .
        .

or

    [if]
        [have_unit]
            type=Horseman
        [/have_unit]
        [then]
            [role]
                role=rider
                type=Horseman
            [/role]
            .
            .
            .

in the case where we need a 'rider' role, but only if the Horseman is already on the board. Yes, it is common for the role simply being used for a message. But how do we KNOW unless we look?

As a systems guy, I'd go for the second because only searches once in the failure case; the first option always searches twice. But, as has been pointed out, there's usually not a lot of data and the processing time is minuscule. So it's more a style question.

CelticMinstrel commented:
That wasn't my commit that removed the message. I think the message is a bit too "in your face" - sometimes it's useful, but sometimes a role really is optional. I'm thinking about implementing the `search_recall_list` and `[else]` suggestions, and in the case of the latter, the preferred method would probably be to use it when a role is not optional.

GregoryLundberg commented:
You're right Zookeeper did that. Sorry.

With his change, I've re-built my PR for Liberty and am about to submit it and close my original PR. This new version is MUCH cleaner with Zookeeper's change. I doubt other campaigns will clean up so easily, but I'll be working on the low-hanging fruit, first, so it will may be a while.

ln-zookeeper commented:
Regarding the above comparison: as a maintenance and compatibility guy, I definitely go for the former, because that's how everything already works right now. Changing it would effectively be a compatibility-breaking API change and leads to everyone having to go through their [role] usage to make sure there's no needless warnings spammed in the user's face.

Sure, when designing things from scratch the latter might be the better choice (although one might argue that tags which operate on a unit don't generally output such errors, and so this one shouldn't either), but that's not the only consideration. Compatibility-breaking changes should always take into account the amount of WML update/maintenance work involved (both mainline and UMC) and actually lead to some kind of substantial improvement that justifies said amount of work. Having [role] output such a warning will of course clearly point out unintended failures in rare situations, but it will also lead to lots and lots of false positives and forces everyone to change their "was there a suitable unit?" checks to happen before instead of after.

The new [role] functionality that was added seems to allow for the current role-assignment problems in scenarios to be nicely fixed in a simpler, less intrusive and completely optional manner.

CelticMinstrel commented:
The warning that zookeeper obtained can now easily be obtained by explicitly placing it in the `[else]`, if that's what someone wants (eg for debugging purposes).

GregoryLundberg commented:
I'm loving both changes.

ln-zookeeper commented:
Are the [role] changes going to lead to this being reworked? I can't really tell what the status of this is.

Anyway, in 18/19c the merman shouldn't be stored/unstored like that. Even if you do refill its movement points, it might still be missing hitpoints and carry status effects, and the simplest way to avoid having to deal with all that is to just let it go to the recall list and then recall it normally in 19c.

GregoryLundberg commented:
Yes. I intend to rework this extensively. I'm learning a lot of the perils of store/unstore digging deep into DM for just those sorts of problems.

It's more than sending to the recall list. You have to let it be on the map at the time [endlevel] hits or it won't get all the status updated. There is a [heal] which handles several status, but the Wiki is worked such that there is no guarantee everything which needs doing is done. And store/unstore is central to the design of DM. Every time I think I have it, I run a slightly different test and I find MORE problems with all that store use.

Anyway, I've been pondering what to do about this PR. I probably should simply close it and submit a new one since it may be a week or three before I get back to it. I've been leaving the PRs I intend to re-work open, though, to keep the history.

CelticMinstrel commented:
I think it's generally better to continue using the same PR rather than always closing them and opening a new one. (Though, you could close it and later reopen the same one, I think.)

GregoryLundberg commented:
Yes to both. Early on I didn't realize GitHub would figure things out so well. Back when I was big into open source we did it all via email and I (the sole maintainer of the CVS) handled all the commits. Shared repositories were still a couple years in the future. This having three repositories takes some getting used to.

GregoryLundberg commented:
I am approaching the end of this. Barring bugs I've not yet spotted, there are only two items left on my list. One, however, I can't decide:

Is it "Scepter of Fire" or "Sceptre of Fire"

My spell checker is en-US and does not like "Sceptre" (which is a valid spelling from Middle English and appeals simply because it is archaic). I'm leaving it alone but raising the question because consistency is probably a good thing.

GregoryLundberg commented:
By count, "Sceptre" is by far more common.

In the character files (which I've never looked at until I grep'd) there are comments saying "re" is intentional and always used.

So, should I change the translatable strings using "er" to "re"?

CelticMinstrel commented:
Your spelling changes are inconsistent. You've changed variable, macro names, and comments to use the American spelling, and you've changed dialogue to use the British spelling. Shouldn't you pick one and stick with it? (Also, why have you capitalized it?)

GregoryLundberg commented:
I changed those few translatable strings to use the capitalized Middle English form, to be consistent with the 100 places it was already that way.

I changes those few program-internal variable names for the same reason. Actually, I think the main change internally was to align the Lua variables to use the same names as the WML variables; and the macro names.

The idea is search for "scepter" and you find the variables (plus the Dark Queen's regular scepter, one occurrence); search for Sceptre and you find the unique item Sceptre of Fire.

GregoryLundberg commented:
BTW, I am about 1/2 way through getting Debug next_level and choose_level working. After that I want to spend some time trying to fix a bug in the map generator (it should not place castles on the bitter edge .. randomly you only get 3 (even only 2) slots to recall/recruit into .. just doesn't seem right, you should get 6 all the time). Barring finding more bugs, that should complete my list ...

GregoryLundberg commented on data/campaigns/Heir_To_The_Throne/utils/abilities.cfg:
```diff
@@ -104,6 +104,12 @@ _"This unit's grasp of melee tactics allows adjacent allies to strike the first
         {NEXT i}
+        [lua] # Can't see why {INITIATIVE_OBJECT} leaves the last unit selected.
+            code=<<
+            wesnoth.deselect_hex()
```
I **_REALLY_** didn't want to do this but I just could not see the issue.

ln-zookeeper replied:
If it happens, it's a bug.

GregoryLundberg replied:
Well, it's 100% reproducible in any of those three scenario (if someone who knows the bowels of the engine wants to got hunting). And it's definitely in the INITIATIVE_OBJECT macro. To reproduce it, you need simply begin the scenario. It only occurs prior to the first turn and effects Li'sar, and troops on her size and, when she and Konrad join forces in S16, Konrad's side, as well.

CelticMinstrel commented:
> The idea is search for "scepter" and you find the variables (plus the Dark Queen's regular scepter, one occurrence); search for Sceptre and you find the unique item Sceptre of Fire.

However, all those variables and macros in fact relate to the Sceptre of Fire, so I find it weird for them to use a different spelling.

GregoryLundberg commented:
**sigh** OK. but it will be a huge patch.  sed is your friend ...

GregoryLundberg commented:
Oh and you'll have to ask again if you want the files renamed as well XP

GregoryLundberg commented:
units/Commander.cfg:        {HTTT_COMMANDER -scepter}
    units/Commander.cfg:        {LEADING_ANIM "units/konrad-commander-scepter-leading.png" "units/konrad-commander-scepter.png" 17,-22}
    units/Commander.cfg:        {DEFENSE_ANIM_FILTERED "units/konrad-commander-scepter-defend.png" "units/konrad-commander-scepter.png" {SOUND_LIST:HUMAN_HIT} (
    units/Commander.cfg:                image="units/konrad-commander-scepter-leading.png:200"
    units/Princess.cfg:        image="units/human-princess-scepter.png"
    units/Lord.cfg:        {HTTT_LORD -scepter}
    units/Lord.cfg:        {LEADING_ANIM "units/konrad-lord-scepter-leading.png" "units/konrad-lord-scepter.png" -16,-21}
    units/Lord.cfg:        {DEFENSE_ANIM_FILTERED "units/konrad-lord-scepter-defend.png" "units/konrad-lord-scepter.png" {SOUND_LIST:HUMAN_HIT} (
    units/Lord.cfg:                image="units/konrad-lord-scepter-leading.png:200"
    utils/httt_utils.cfg:        mask="{campaigns/Heir_To_The_Throne/masks/17_Scepter_of_Fire.mask}"

    units/Commander.cfg:#define HTTT_COMMANDER AFFIX
    units/Commander.cfg:    image="units/konrad-commander{AFFIX}.png"
    units/Commander.cfg:    {DEFENSE_ANIM_RANGE "units/konrad-commander{AFFIX}-defend.png" "units/konrad-commander{AFFIX}.png" {SOUND_LIST:HUMAN_HIT} melee}
    units/Commander.cfg:            image="units/konrad-commander{AFFIX}.png:25"
    units/Commander.cfg:            image="units/konrad-commander{AFFIX}-attack-[1~4].png:[75,150,75*2]"
    units/Lord.cfg:#define HTTT_LORD AFFIX
    units/Lord.cfg:    image="units/konrad-lord{AFFIX}.png"
    units/Lord.cfg:    {DEFENSE_ANIM_RANGE "units/konrad-lord{AFFIX}-defend.png" "units/konrad-lord{AFFIX}-attack-w1.png" {SOUND_LIST:HUMAN_HIT} melee}
    units/Lord.cfg:            image="units/konrad-lord{AFFIX}-attack-w[1,3,1].png:[100,200,75]"
    units/Lord.cfg:            image="units/konrad-lord{AFFIX}-attack-w[1,2,1].png:[100,200,75]"

    -rw-r--r-- 1 lundberg lundberg 2166 Jun 29 21:08 human-battleprincess-scepter-attack-1.png
    -rw-r--r-- 1 lundberg lundberg 4722 Jun 29 21:08 human-battleprincess-scepter-attack-2.png
    -rw-r--r-- 1 lundberg lundberg 2282 Jun 29 21:08 human-battleprincess-scepter-defend-1.png
    -rw-r--r-- 1 lundberg lundberg 2138 Jun 29 21:08 human-battleprincess-scepter-defend-2.png
    -rw-r--r-- 1 lundberg lundberg 1958 Jun 29 21:08 human-battleprincess-scepter-leading.png
    -rw-r--r-- 1 lundberg lundberg 2212 Jun 29 21:08 human-battleprincess-scepter.png
    -rw-r--r-- 1 lundberg lundberg 1522 Jun 29 21:08 human-princess-scepter-defend.png
    -rw-r--r-- 1 lundberg lundberg 1859 Jun 29 21:08 human-princess-scepter-leading.png
    -rw-r--r-- 1 lundberg lundberg 1509 Jun 29 21:08 human-princess-scepter.png
    -rw-r--r-- 1 lundberg lundberg 2708 Jun 29 21:08 konrad-commander-scepter-attack-1.png
    -rw-r--r-- 1 lundberg lundberg 3549 Jun 29 21:08 konrad-commander-scepter-attack-2.png
    -rw-r--r-- 1 lundberg lundberg 2429 Jun 29 21:08 konrad-commander-scepter-attack-3.png
    -rw-r--r-- 1 lundberg lundberg 2398 Jun 29 21:08 konrad-commander-scepter-attack-4.png
    -rw-r--r-- 1 lundberg lundberg 3026 Jun 29 21:08 konrad-commander-scepter-defend.png
    -rw-r--r-- 1 lundberg lundberg 2335 Jun 29 21:08 konrad-commander-scepter-leading.png
    -rw-r--r-- 1 lundberg lundberg 1791 Jun 29 21:08 konrad-commander-scepter.png
    -rw-r--r-- 1 lundberg lundberg 1782 Jun 29 21:08 konrad-lord-scepter-attack-w1.png
    -rw-r--r-- 1 lundberg lundberg 3897 Jun 29 21:08 konrad-lord-scepter-attack-w2.png
    -rw-r--r-- 1 lundberg lundberg 3690 Jun 29 21:08 konrad-lord-scepter-attack-w3.png
    -rw-r--r-- 1 lundberg lundberg 1936 Jun 29 21:08 konrad-lord-scepter-defend.png
    -rw-r--r-- 1 lundberg lundberg 2141 Jun 29 21:08 konrad-lord-scepter-leading.png
    -rw-r--r-- 1 lundberg lundberg 1773 Jun 29 21:08 konrad-lord-scepter.png
    -rw-r--r-- 1 lundberg lundberg 1809 Jun 29 21:08 konrad-lord-scepter-sword-defend.png
    -rw-r--r-- 1 lundberg lundberg 14211 Jun 29 21:08 17_Scepter_of_Fire.mask
    -rw-r--r-- 1 lundberg lundberg 24430 Aug  2 15:12 17_Scepter_of_Fire.cfg

GregoryLundberg commented:
OK, after a quick run-through, it does not appear anything broke with all that renaming. Profiles change as they should, etc. No errors or warnings on screen or stderr. I did a quick check of the animations and they look correct .. but I might have missed something there .. they're small and it's are to tell how many frames you're seeing.

CelticMinstrel commented:
Let's not worry about the filenames.

ln-zookeeper commented:
The (current) use of sceptre/scepter in the text seems pretty consistent to me. The common noun is "scepter", and the name of the artifact is "Sceptre of Fire"; the artifact can naturally be referred to using either, but never by "Scepter". That was deliberately standardized in the text years ago, and I see no reason to assume there's anything wrong with it now.

One could rename variables and the scenario/map files, but I rather fail to see the benefit in that.

CelticMinstrel commented:
Like zookeeper said, I don't see any benefit in renaming variables or macros in either direction, though I won't deny that it's nice to be consistent.

On "scepter" vs "Sceptre" I have no real opinion, I guess. (If I had my way it'd be sceptre everywhere. :p Maybe I should just play the en-gb version, though that'd be odd in other ways.(

GregoryLundberg commented:
There were a _few_ instances where the proper object Sceptre was referred to as scepter (lower case) in the translatable text. So some where missed or changed after standardization.

The only general scepter is the Dark Queen's scepter weapon/attack.

The internal variable names where scepter except in Lua. But they don't really matter. Consistency, however, is a Good Thing (sm).

gfgtdf commented on data/campaigns/Heir_To_The_Throne/scenarios/11_The_Ford_of_Abez.cfg:
```diff
@@ -299,47 +298,84 @@
+                x,y=$x1,$y1
+                [and]
+                    x,y=5,27
+                    radius=11
```
Did you test that it still performs well? Note that radius= can be quite slow, i know for sure at least one place where the code was changed from radius= to use xy ranges becasue of performance issues. I cannot tell without testing whether this is important here though.

gfgtdf commented:
about a73bb7a and f5bad69, i'd be better to fix the underlying isue than to fix the wml code here, unfortunateley i wasn't able to reprporudce them, please file a bug report in the bugtracker and provide the folowing information:
1) Did you use a 32 or a 64 bit build of wesnorth?
2) Did you use debug commands to end the scenario, if so, which debug coimmands did you use?
3) Provide a start-of-scenario save for the scenario where the error message appeared.

GregoryLundberg commented:
Don't have an account and so totally uninterested in creating yet-another-useless-web-account to be forgotten about. I'm synced to master and rebuilt overnight. I'll checkout master and retest but it was 100% reproducible, here. Perhaps that's changed.

GregoryLundberg commented:
[lundberg@Arch wesnoth]$ git checkout master
    Switched to branch 'master'
    Your branch is up-to-date with 'origin/master'.
    [lundberg@Arch wesnoth]$ git log -1
    commit f321e716853c6bd632f4bf2a8672f907257521ed
    Author: fendrin <fendrin@gmx.de>
    Date:   Thu Aug 4 03:09:30 2016 +0200
    
        SoF 08: Removed some more [+unit].
    [lundberg@Arch wesnoth]$ ~/wesnoth/wesnoth
    Battle for Wesnoth v1.13.5+dev (f321e71-Clean)
    Started on Thu Aug  4 10:09:59 2016
    
    Automatically found a possible data directory at /home/lundberg/wesnoth
    
    Data directory:               /home/lundberg/wesnoth
    User configuration directory: /home/lundberg/.config/wesnoth
    User data directory:          /home/lundberg/.local/share/wesnoth/1.13
    Cache directory:              /home/lundberg/.cache/wesnoth
    
    Setting mode to 1440x851
    Checking lua scripts... ok
    [New campaign, HttT, Champion]
    [S01 moves=99, move Konrad to goalpost]
    [S02 kill enemy shift-k]
    [S03 kill enemy shift-k]
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    libpng warning: iCCP: known incorrect sRGB profile
    [S04 kill enemies, shift-k twice]
    [Choose by land]
    [S05a moves=99, move Konrad to end of road]
    [S06 kill enemies, shift-k twice]
    20160804 10:18:34 error engine: Trying to add Kalenz - Kalenz - 2 (3,12) over Delfador - Delfador - 2 (3,11).
    20160804 10:18:34 error engine: The new unit was assigned underlying_id=63 to prevent duplicate id conflicts.
    [S07 have begun, prestart run, ending test once inital messages and goals displayed]
    [lundberg@Arch wesnoth]$

gfgtdf commented:
coudl you please try with lastest master with 531b91359a0fb465883839bffbe51f29e35d56c7 in ?

Also:
> 2) Did you use debug commands to end the scenario, if so, which debug coimmands did you use?

If you don't want to make a gna account you can also post it here

GregoryLundberg commented:
Bug is in S06. Errors appear at start of S07. I can give you a scenario-start for S06 but you'll have to step through anyway. Scenario-start for S07 will simply show it was bugged in S06.

Arch Linux, 64-bit VM under VirtualBox on Windows 10 host connected via Windows remote desktop.

gfgtdf commented:
Thx, im pretty sure that here ar 2 bugs involved: one about 'debug-kill' beeing unsynced and one about underlying unit ids corrupting on 64bit systems when they are too high.

I'd like to fix bug #2 first, and since i don't run a 64 bit system i cannot test it, i'm hoping to have that fixed with 531b91359a so i'd like you to test it again with that commit if it is fixed you might still get a error message but with much higher human than just '2' so mabye somthign like `error engine: Trying to add Kalenz - Kalenz - 9223372036854775810 (3,12) over Elvish Shaman - Elvish Shaman - 9223372036854775810 (3,11).` To reprporidce this you migth need to acctivate plannng mode (press p) and recruit 2 unit in planning mode in that scenario 7.

GregoryLundberg commented:
OK. Retested.

'throw enemies defeated' --> no message
shift-k on two enemies --> message
create Ancient Lich next to enemies and attack normally --> no message

All using the scenario-start for S06 created in the earlier post. So it's not the [unit] tag doing it, it's caused by how you got to 'enemies defeated'

GregoryLundberg commented:
BTW For my PR, this is a moot issue. The cleanup needed for debug next_level and debug choose_level revealed a design flaw in saving and restoring units and getting them working cleanly meant Kalenz and Li'sar are always created in prestart anyway.

GregoryLundberg commented:
If you're looking for a bug to swat which matters, check the line comment about about the lua whack to deselect hex. Or check out my issue https://github.com/GregoryLundberg/wesnoth/issues/17

gfgtdf commented:
Hmm no, i really mostly care about this bug about bad invalid unit ids, even if you work around it in this camapign there might stil be multiple other ampaign out there that suffer from it.

GregoryLundberg commented:
OK. sync'd and made ...

    [lundberg@Arch wesnoth]$ git checkout master
    Already on 'master'
    Your branch is up-to-date with 'origin/master'.
    [lundberg@Arch wesnoth]$ git log -3 | tail -5
    commit 531b91359a0fb465883839bffbe51f29e35d56c7
    Author: gfgtdf <daniel.gfgtdf@gmail.com>
    Date:   Thu Aug 4 15:59:56 2016 +0200
    
        fix wrong underlying_id parsing on 64 bit systems.
    [lundberg@Arch wesnoth]$ ~/wesnoth/wesnoth
    Battle for Wesnoth v1.13.5+dev (e75325a-Clean)
    Started on Thu Aug  4 11:29:12 2016

Using the scenario-start from before, and shift-k to kill the leaders and the messages no longer appear.
Checking Kalenz' with inspect shows underlying_id=9223372036854775810

I'd be concerned that what this does is make the saved turns, replays, etc, unusable on 32-bit systems. Unfortunately, I don't have a 32-bit system to test on :)

GregoryLundberg commented:
Oh, yea, DM suffers from this bug BIG TIME. I just declared it not-my-problem and ignored it, then.

gfgtdf commented:
> I'd be concerned that what this does is make the saved turns, replays, etc, unusable on 32-bit systems. Unfortunately, I don't have a 32-bit system to test on :) 

Hmm maybe we shoudl just make that some uint32_t type instead of size_t, well i still sonsider this an imvovement over the previous code where it always cut of the first bits in 64 bit modes. I'll now try to fix the debug-kill issue.

GregoryLundberg commented:
uh huh. size_t is non-portable and poorly spec'd I avoid it whenever possible.

GregoryLundberg commented:
We're in C++ and I'd encapsulate unit ID and such and avoid having to look for all the places where a bad data type was used. That way an issue like this has a single-point-of-change. Instead you're going to have to grep everywhere and hope you get 'em all.

GregoryLundberg commented:
[Note to self: create an issue about using local date/time being such a bad idea and looking where all it's used and upgrading to using UTC so we're all on the same page.]

gfgtdf commented:
ok i pushed a new commit a0233a1d4df6463323b06bdf3f32f3b1345cf574 to fix (sync) debug-kill commands

GregoryLundberg commented:
OK, restarted using the process above since I was doing other stuff in the interim ..

inspect at end of S06 shows Kalenz' underlying_id=29
no message on stderr
inspect at start of S07 shows still =29

Looks good to me. 29 sounds about right counting units in my head to the point he gets created.

GregoryLundberg commented:
Closing in favor of new PRs
